x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/GrokEvaluatorExtracter.java

```diff
@@ -42,62 +45,67 @@ public GrokEvaluatorExtracter(
         this.parser = parser;
         this.pattern = pattern;
         this.valuesSet = new boolean[types.size()];
+        this.firstValues = new Object[types.size()];
+        this.positionToType = new ElementType[types.size()];
+
         fieldExtracters = new ArrayList<>(parser.captureConfig().size());
         for (GrokCaptureConfig config : parser.captureConfig()) {
+            var key = config.name();
+            ElementType type = types.get(key);
+            Integer blockIdx = keyToBlock.get(key);
+            positionToType[blockIdx] = type;
+
             fieldExtracters.add(config.objectExtracter(value -> {
```

Comment:
Probably unrelated to this PR, but if you build the extracter via `nativeExtracter` instead of `objectExtracter` you could skip the boxing in most cases.

Reply:
And you wouldn't need `BlockUtils.appendValue(block, value, type);` and just call the append directly. It's much more code but cuts down the number of allocations and switches.

Reply:
And would avoid the `instanceOf Float` - you'd know you are in the `float` path and handle it natively.